### PR TITLE
WT-11344 Make background compaction interruptible

### PR DESCRIPTION
**Pull request by etienneptl**
Background compaction follows the same behaviour as foreground compaction when it is interrupted:

- `ETIMEDOUT` if the timer has elapsed
- `WT_ERROR` if the server has been turned off while compaction was happening

### CODE REVIEW

File: test/suite/test_compact07.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+import time
+import wiredtiger, wttest
+
+# test_compact07.py
+# Test background compaction interruption.
+class test_compact07(wttest.WiredTigerTestCase):
+    uri = 'file:test_compact07'
```
Comment (Sean04) [inline test/suite/test_compact07.py:35]: I don't think this uri is used

File: test/suite/test_compact07.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+        # Make sure compaction will start working by setting a low threshold.
+        compact_cfg_on = 'background=true,free_space_target=1MB'
+        compact_cfg_off = 'background=false'
+        with self.expectedStderrPattern('background compact interrupted by application'):
+            self.session.compact(None, compact_cfg_on)
+            # Give the background compaction server some time to wake up.
+            time.sleep(1)
+            # Interrupt it.
+            self.session.compact(None, compact_cfg_off)
```
Comment (Sean04) [inline test/suite/test_compact07.py:85]: We should check the background compact server status stats after this.

File: test/suite/test_compact07.py (hunk 1 above)
Comment (etienneptl) [reply test/suite/test_compact07.py]: Removed in b2dfd29d8a2a2c16456723c3b86affe7488565f7

File: test/suite/test_compact07.py (hunk 2 above)
Comment (etienneptl) [reply test/suite/test_compact07.py]: Good idea, checking this in e898ddfbc6b65a1d4a2e6f5185905846035fde7c.

**Review by Sean04 — APPROVED** [review]
LGTM, after adding the stat to track and test when the background server has been interrupted.

**Comment by etienneptl** [comment]
The current failures are showing the expected log `background compact interrupted by application`. I will find a way to ignore this message and will add a stat to test the background compaction server has been interrupted as expected.

**Comment by etienneptl** [comment]
> The current failures are showing the expected log `background compact interrupted by application`. I will find a way to ignore this message and will add a stat to test the background compaction server has been interrupted as expected.

I managed to use both in the end, see 4f586d382a1bd86724141ea3cd521cf442b58f57. I was facing a timing issue since the beginning. The `while` loops waiting for the server to be on/off was the key.

File: test/suite/test_compact08.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                # working.
+                self.session.compact(None, 'background=true,free_space_target=1MB')
+
+                # Wait for the server to wake up.
+                running = self.get_bg_compaction_running()
+                while not running:
+                    running = self.get_bg_compaction_running()
+                self.assertEqual(running, 1)
+
+                # Disable the server which may interrupt compaction.
```
Comment (kommiharibabu) [inline test/suite/test_compact08.py:110]: I am not sure that in slow machines where the background compact server can get a chance to start compaction on a table before we disable it. How about adding 1 second sleep here to increase the probability of bg compact starts the compaction?

Comment (etienneptl) [reply test/suite/test_compact08.py:129]: Good idea, see 2fa8221474751f22a9d2bca27d1d0e86820ec755 where I am retrieving the stat `btree_compact_pages_rewritten` to make sure we are doing work.

**Review by kommiharibabu — APPROVED** [review]
LGTM

File: test/suite/test_compact08.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                    running = self.get_bg_compaction_running()
+                self.assertEqual(running, 1)
+
+                # Wait for the server to start compacting.
+                num_pages_rewritten = 0
+                while num_pages_rewritten == 0:
+                    for j in range(self.num_tables):
+                        uri = f'{self.table_uri}_{j}'
+                        num_pages_rewritten = self.get_pages_rewritten(uri)
+                        self.prout(num_pages_rewritten)
```
Comment (etienneptl) [inline test/suite/test_compact08.py:124]: This should be removed

**Comment by etienneptl** [comment]
I reverted [2fa8221](https://github.com/wiredtiger/wiredtiger/pull/9485/commits/2fa8221474751f22a9d2bca27d1d0e86820ec755) as the stat is incremented too late and the interruption cannot be detected.

**Comment by etienneptl** [comment]
evergreen merge

**Pull request merged**